File: web/src/types/office.ts
/**
 * My Office contracts — mirrors backend/app/models/office.py field-for-field.
 *
 * The evidence projection is the USER layer of trace-observability.md: it
 * never carries latency, usage, cost, model/provider names, confidence
 * values, or raw tool payloads.
 */

export type OfficeSessionStatus = "active" | "archived";

export type OfficeSession = {
  id: string;
  user_id: string;
  title: string;
  status: OfficeSessionStatus;
  last_message_preview: string;
  created_at: string;
  updated_at: string;
};

export type OfficeMessage = {
  id: string;
  role: string;
  content: string;
  request_id?: string | null;
  execution?: TurnExecutionFacts | null;
  created_at: string;
};

export type TurnExecutionFacts = {
  outcome:
    | "direct_response"
    | "clarification"
    | "executed"
    | "blocked"
    | "failed";
  evidence_available: boolean;
  specialist_findings_available: boolean;
  process_available: boolean;
  policy_blocked: boolean;
};

export type EvidenceStepKind = "tool" | "specialist" | "audit";

export type EvidenceStep = {
  label: string;
  kind: EvidenceStepKind;
  done: boolean;
};

export type EvidenceFinding = {
  agent: string;
  title: string;
  evidence: string[];
};

export type DataCoverage = {
  period?: string | null;
  source_counts: Record<string, number>;
  quality_warnings: string[];
};

export type EvidenceAudit = {
  status: string;
  warnings: string[];
};

export type AdvancedDetails = {
  run_id: string;
  tool_names: string[];
};

export type UserEvidenceProjection = {
  request_id: string;
  findings: EvidenceFinding[];
  cited_sources: string[];
  data_coverage: DataCoverage;
  steps: EvidenceStep[];
  audit?: EvidenceAudit | null;
  advanced?: AdvancedDetails | null;
};

File: backend/app/models/office.py
"""My Office contracts: sessions, messages, and the user evidence projection.

The evidence projection is the USER layer of `trace-observability.md` — it
must never carry latency, usage, cost, model/provider names, confidence
values, or raw tool payloads. Those belong to the developer projection only.
"""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.turn_execution import TurnExecutionFacts


class OfficeSession(BaseModel):
    id: str
    user_id: str
    title: str = ""
    status: Literal["active", "archived"] = "active"
    last_message_preview: str = ""
    created_at: str
    updated_at: str


class OfficeMessage(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    role: str
    content: str
    request_id: Optional[str] = None
    execution: Optional[TurnExecutionFacts] = None
    created_at: str


class CreateSessionRequest(BaseModel):
    user_id: str = "demo"
    title: str = ""


class UpdateSessionRequest(BaseModel):
    title: Optional[str] = None
    status: Optional[Literal["active", "archived"]] = None


class EvidenceStep(BaseModel):
    label: str
    kind: Literal["tool", "specialist", "audit"]
    done: bool = True


class EvidenceFinding(BaseModel):
    agent: str
    title: str
    evidence: list[str] = Field(default_factory=list)


class DataCoverage(BaseModel):
    period: Optional[str] = None
    source_counts: dict[str, int] = Field(default_factory=dict)
    quality_warnings: list[str] = Field(default_factory=list)


class EvidenceAudit(BaseModel):
    status: str
    warnings: list[str] = Field(default_factory=list)


class AdvancedDetails(BaseModel):
    run_id: str
    tool_names: list[str] = Field(default_factory=list)


class UserEvidenceProjection(BaseModel):
    request_id: str
    findings: list[EvidenceFinding] = Field(default_factory=list)
    cited_sources: list[str] = Field(default_factory=list)
    data_coverage: DataCoverage = Field(default_factory=DataCoverage)
    steps: list[EvidenceStep] = Field(default_factory=list)
    audit: Optional[EvidenceAudit] = None
    advanced: Optional[AdvancedDetails] = None
